File: src/aw_index_cli/compose.py
"""Select repositories from a distribution and render a vcs ``.repos`` file."""

from __future__ import annotations

import yaml


class ComposeError(Exception):
    """Raised when a distribution cannot be composed into ``.repos`` entries."""


def _reject_unknown(singular: str, plural: str, missing: set[str]) -> None:
    """Raise :class:`ComposeError` naming any explicitly-requested unknowns."""
    if missing:
        names = ", ".join(repr(name) for name in sorted(missing))
        label = singular if len(missing) == 1 else plural
        raise ComposeError(f"no such {label} in the distribution: {names}")


def select_repositories(
    distribution: dict,
    tags: list[str] | None = None,
    *,
    packages: list[str] | None = None,
    repository: list[str] | None = None,
    include_dependencies: bool = True,
    reference_design: bool = False,
) -> list[tuple[str, dict, list[str]]]:
    """Return ``(key, spec, selected_packages)`` triples sorted by repo key.

    Four optional filters narrow the selection and are ANDed together; omit
    all of them to select the whole distribution:

    * ``tags``: keep packages whose own ``tags`` intersect these.
    * ``packages``: keep packages whose name is in this list.
    * ``repository``: keep only these repository entries (by registry key).
    * ``reference_design``: keep only entries marked as a reference design.

    A repository is selected when at least one of its packages survives every
    given filter. For a v4 distribution, the selected packages then pull in
    their transitive ``index_dependencies``, even when those dependencies do
    not match the filters. ``include_dependencies=False`` keeps the original
    filtered selection for commands such as ``list``. Package names within
    each repository are sorted. An explicit
    ``repository`` key or ``packages`` name that is absent from the *whole*
    distribution (independent of the other filters, so a typo never hides
    behind an empty result) raises :class:`ComposeError`, as does a selected
    repository whose ``packages`` is not a mapping.
    """
    all_repos = distribution.get("repositories") or {}
    wanted_tags = set(tags or [])
    wanted_pkgs = set(packages or [])
    wanted_repos = set(repository or [])

    # Validate explicit names against the entire distribution before filtering,
    # so an unknown name errors loudly rather than yielding silent-empty output.
    known_pkgs: set[str] = set()
    for spec in all_repos.values():
        spec_pkgs = (spec or {}).get("packages")
        if isinstance(spec_pkgs, dict):
            known_pkgs.update(spec_pkgs)
    _reject_unknown("repository entry", "repository entries", wanted_repos - set(all_repos))
    _reject_unknown("package", "packages", wanted_pkgs - known_pkgs)

    selected = []
    for key, spec in sorted(all_repos.items()):
        if wanted_repos and key not in wanted_repos:
            continue
        marker = (spec or {}).get("reference_design")
        if marker is not None and not isinstance(marker, bool):
            raise ComposeError(
                f"repository {key!r} has 'reference_design' that is not a boolean "
                f"(got {type(marker).__name__})"
            )
        if reference_design and not marker:
            continue
        spec_pkgs = (spec or {}).get("packages") or {}
        if not isinstance(spec_pkgs, dict):
            raise ComposeError(
                f"repository {key!r} has 'packages' that is not a mapping "
                f"of package name to spec (got {type(spec_pkgs).__name__})"
            )
        names = sorted(
            name
            for name, pkg in spec_pkgs.items()
            if (not wanted_tags or set((pkg or {}).get("tags") or []) & wanted_tags)
            and (not wanted_pkgs or name in wanted_pkgs)
        )
        if names:
            selected.append((key, spec, names))
    if include_dependencies and distribution.get("schema_version") == "4" and selected:
        return _with_index_dependencies(all_repos, selected)
    return selected


def _with_index_dependencies(
    all_repos: dict, selected: list[tuple[str, dict, list[str]]]
) -> list[tuple[str, dict, list[str]]]:
    """Expand filtered roots to their full v4 package dependency closure."""
    owners: dict[str, tuple[str, dict]] = {}
    for repo_key, repo in sorted(all_repos.items()):
        package_specs = (repo or {}).get("packages") or {}
        if not isinstance(package_specs, dict):
            continue  # A selected malformed repository already fails above.
        for name, package_spec in package_specs.items():
            if name in owners:
                raise ComposeError(f"package {name!r} is registered by multiple repositories")
            owners[name] = (repo_key, package_spec if package_spec is not None else {})

    included = {name for _key, _spec, names in selected for name in names}
    visited: set[str] = set()
    visiting: list[str] = []

    def visit(name: str) -> None:
        if name in visiting:
            cycle = visiting[visiting.index(name) :] + [name]
            raise ComposeError(f"index dependency cycle: {' -> '.join(cycle)}")
        if name in visited:
            return
        repo_key, package_spec = owners[name]
        if not isinstance(package_spec, dict):
            raise ComposeError(f"package {name!r} in repository {repo_key!r} is not a mapping")
        dependencies = package_spec.get("index_dependencies", [])
        if not isinstance(dependencies, list) or any(
            not isinstance(dependency, str) for dependency in dependencies
        ):
            raise ComposeError(
                f"package {name!r} has invalid 'index_dependencies' "
                "(expected a list of package names)"
            )
        visiting.append(name)
        for dependency in sorted(dependencies):
            if dependency not in owners:
                raise ComposeError(
                    f"index dependency {dependency!r} of package {name!r} is not registered"
                )
            included.add(dependency)
            visit(dependency)
        visiting.pop()
        visited.add(name)

    for name in sorted(included):
        visit(name)

    by_repo: dict[str, list[str]] = {}
    for name in included:
        repo_key, _package_spec = owners[name]
        by_repo.setdefault(repo_key, []).append(name)
    return [(key, all_repos[key], sorted(names)) for key, names in sorted(by_repo.items())]


def unknown_tags(distribution: dict, tags: list[str] | None) -> list[str]:
    """Return the requested tags that no package in the distribution carries.

    Sorted for stable output. Unlike an unknown ``--packages`` name or
    ``--repository`` key, an unknown tag is not a hard error: the tag
    vocabulary lives in the registry, not in the distribution file, so a
    valid id may simply have no usage yet. The CLI reports these as a
    stderr warning while still composing the (possibly empty) output.
    """
    if not tags:
        return []
    carried: set[str] = set()
    for spec in (distribution.get("repositories") or {}).values():
        spec_pkgs = (spec or {}).get("packages")
        if not isinstance(spec_pkgs, dict):
            continue
        for pkg in spec_pkgs.values():
            carried.update((pkg or {}).get("tags") or [])
    return sorted(set(tags) - carried)


def to_repos_entries(repositories: list[tuple[str, dict, list[str]]]) -> dict:
    """Map selected repositories to an ordered ``key -> entry`` dict.

    The entry key is the registry repository key, so packages from one
    monorepo collapse into a single clone. Each entry carries exactly
    vcs2l's ``type``/``url``/``version`` and nothing else; the format
    defines no other per-entry fields. The selected registered package names
    are recorded in the provenance header comments (see
    :func:`provenance_header`), not in the YAML body. Raises
    :class:`ComposeError` when a repository is missing ``url`` or
    ``ref.value``, or when its ``ref`` is not a mapping.
    """
    entries: dict = {}
    for key, spec, _names in repositories:
        spec = spec or {}
        url = spec.get("url")
        if not url:
            raise ComposeError(f"repository {key!r} is missing 'url'")
        ref = spec.get("ref") or {}
        if not isinstance(ref, dict):
            raise ComposeError(
                f"repository {key!r} has 'ref' that is not a mapping "
                f"with 'kind' and 'value' (got {type(ref).__name__})"
            )
        version = ref.get("value")
        if not version:
            raise ComposeError(f"repository {key!r} is missing 'ref.value'")
        entries[key] = {
            "type": "git",
            "url": url,
            "version": version,
        }
    return entries


def provenance_header(
    *,
    tool_version: str,
    ros_distro: str,
    source: str,
    tags: list[str] | None = None,
    packages: list[str] | None = None,
    repository: list[str] | None = None,
    reference_design: bool = False,
    autoware: str | None = None,
    generated_at: str | None = None,
    selection: list[tuple[str, list[str]]] | None = None,
) -> list[str]:
    """Build the ``# …`` comment lines that precede the rendered ``.repos``.

    The ``packages`` and ``repository`` selection filters, when given, are
    recorded so the file documents how it was produced. ``selection`` is the
    ``(repo_key, selected_package_names)`` listing; when given, every entry is
    named in the header with its selected packages.
    """
    lines = [
        f"# aw-index-cli {tool_version}",
        f"# source: {source}",
        f"# rosdistro: {ros_distro}",
        f"# tags: {', '.join(tags) if tags else 'all'}",
    ]
    if packages:
        lines.append(f"# packages: {', '.join(packages)}")
    if repository:
        lines.append(f"# repository: {', '.join(repository)}")
    if reference_design:
        lines.append("# reference_design: true")
    if autoware is not None:
        lines.append(
            f"# autoware: {autoware} "
            "(informational only, not a ref selector; the registry tracks "
            "one ref per repository)"
        )
    if generated_at is not None:
        lines.append(f"# generated_at: {generated_at}")
    if selection:
        lines.append("# selected packages by repository:")
        for key, package_names in selection:
            lines.append(f"#   {key}: {', '.join(package_names)}")
    lines.append(
        "# Generated file. Re-run 'aw-index-cli compose …' to update; " "do not edit by hand."
    )
    return lines


def render_repos(
    distribution: dict,
    *,
    tags: list[str] | None = None,
    packages: list[str] | None = None,
    repository: list[str] | None = None,
    reference_design: bool = False,
    header_lines: list[str],
) -> str:
    """Render the full ``.repos`` document (header comments + YAML body).

    A blank line separates the ``#`` provenance header from the YAML body so the
    two read as distinct sections.
    """
    repositories = select_repositories(
        distribution,
        tags=tags,
        packages=packages,
        repository=repository,
        reference_design=reference_design,
    )
    entries = to_repos_entries(repositories)
    body = yaml.safe_dump(
        {"repositories": entries},
        sort_keys=False,
        default_flow_style=False,
    )
    return "\n".join(header_lines) + "\n\n" + body

File: js/compose.mjs
// aw-index-cli — canonical registry→`.repos` composition, as an ES module.
//
// This mirrors `src/aw_index_cli/compose.py` 1:1 so a JavaScript consumer (the
// autoware-index browse site's "repos builder") can reuse the SAME transform
// instead of forking a second, drifting implementation. The Python CLI remains
// the reference; `tests/test_conformance.py` runs both this module and the
// Python compose over shared fixtures and asserts identical `.repos` *content*
// (same parsed body + same header lines), so the two cannot silently diverge.
// Exact byte formatting (scalar quoting style, line wrapping) need not match.
//
// Dependency-free on purpose: no npm, no bundler. It is importable directly in
// the browser via `<script type="module">` and by Node's built-in test runner.

/** Keep equal to `src/aw_index_cli/__init__.py` `__version__` (asserted by conformance). */
export const VERSION = "0.4.0";

/** Registry defaults, mirroring `registry.py` `DEFAULT_REPO`/`DEFAULT_REF`. */
export const DEFAULT_REPO = "autowarefoundation/autoware-index";
export const DEFAULT_REF = "main";

/** Raised when a distribution cannot be composed into `.repos` entries. */
export class ComposeError extends Error {
  constructor(message) {
    super(message);
    this.name = "ComposeError";
  }
}

// Sort by code point, matching Python's `sorted()` on ASCII registry keys/names.
const cmp = (a, b) => (a < b ? -1 : a > b ? 1 : 0);

const isMapping = (v) => v !== null && typeof v === "object" && !Array.isArray(v);

// Mirror Python's `x or {}` for a `packages`/`ref` value: any empty or absent
// container (`null`/`undefined`/`[]`/`{}`) becomes an empty mapping, so an empty
// `packages: []`/`ref: []` selects nothing (Python) rather than raising. JS `[]`
// and `{}` are truthy, so `|| {}` alone would not match Python here.
const mappingOrEmpty = (v) => {
  if (!v) return {};
  if (Array.isArray(v)) return v.length ? v : {};
  return v;
};

function rejectUnknown(singular, plural, missing) {
  if (missing.length) {
    const names = missing
      .slice()
      .sort(cmp)
      .map((n) => `'${n}'`)
      .join(", ");
    const label = missing.length === 1 ? singular : plural;
    throw new ComposeError(`no such ${label} in the distribution: ${names}`);
  }
}

/**
 * Return `[key, spec, selectedNames]` triples sorted by repo key.
 *
 * The three optional filters (`tags`, `packages`, `repository`) are ANDed;
 * omit all to select the whole distribution. An explicit `repository` key or
 * `packages` name absent from the *whole* distribution throws `ComposeError`
 * (so a typo never hides behind an empty result). Mirrors
 * `compose.select_repositories`.
 */
export function selectRepositories(
  distribution,
  { tags = null, packages = null, repository = null } = {},
) {
  const allRepos = (distribution && distribution.repositories) || {};
  const wantedTags = new Set(tags || []);
  const wantedPkgs = new Set(packages || []);
  const wantedRepos = new Set(repository || []);

  const knownPkgs = new Set();
  for (const spec of Object.values(allRepos)) {
    const specPkgs = spec && spec.packages;
    if (isMapping(specPkgs)) {
      for (const name of Object.keys(specPkgs)) knownPkgs.add(name);
    }
  }
  rejectUnknown(
    "repository entry",
    "repository entries",
    [...wantedRepos].filter((r) => !Object.hasOwn(allRepos, r)),
  );
  rejectUnknown(
    "package",
    "packages",
    [...wantedPkgs].filter((p) => !knownPkgs.has(p)),
  );

  const selected = [];
  for (const key of Object.keys(allRepos).sort(cmp)) {
    if (wantedRepos.size && !wantedRepos.has(key)) continue;
    const spec = allRepos[key] || {};
    const specPkgs = mappingOrEmpty(spec.packages);
    if (!isMapping(specPkgs)) {
      throw new ComposeError(
        `repository '${key}' has 'packages' that is not a mapping of package name to spec`,
      );
    }
    const names = Object.keys(specPkgs)
      .filter((name) => {
        const pkg = specPkgs[name] || {};
        const pkgTags = new Set(pkg.tags || []);
        const tagOk = wantedTags.size === 0 || [...pkgTags].some((t) => wantedTags.has(t));
        const pkgOk = wantedPkgs.size === 0 || wantedPkgs.has(name);
        return tagOk && pkgOk;
      })
      .sort(cmp);
    if (names.length) selected.push([key, spec, names]);
  }
  return selected;
}

/**
 * Map selected repositories to an ordered `Map<key, {type,url,version}>`.
 *
 * The entry key is the registry repository key, so a monorepo's packages
 * collapse into one clone. Each entry carries exactly vcs2l's
 * `type`/`url`/`version`; `version` is `ref.value` verbatim. Throws
 * `ComposeError` on missing `url`/`ref.value` or a non-mapping `ref`. Mirrors
 * `compose.to_repos_entries` (a Map preserves insertion order for every key).
 */
export function toReposEntries(repositories) {
  const entries = new Map();
  for (const [key, specRaw] of repositories) {
    const spec = specRaw || {};
    const url = spec.url;
    if (!url) throw new ComposeError(`repository '${key}' is missing 'url'`);
    const ref = mappingOrEmpty(spec.ref);
    if (!isMapping(ref)) {
      throw new ComposeError(
        `repository '${key}' has 'ref' that is not a mapping with 'kind' and 'value'`,
      );
    }
    const version = ref.value;
    if (!version) throw new ComposeError(`repository '${key}' is missing 'ref.value'`);
    entries.set(key, { type: "git", url, version });
  }
  return entries;
}

/**
 * Build the `# …` comment lines that precede the rendered `.repos`.
 * Mirrors `compose.provenance_header` — same order and the same conditional
 * gates (`autoware`/`generatedAt` on `!= null`; the rest on truthiness).
 */
export function provenanceHeader({
  toolVersion,
  rosDistro,
  source,
  tags = null,
  packages = null,
  repository = null,
  autoware = null,
  generatedAt = null,
  selection = null,
} = {}) {
  const lines = [
    `# aw-index-cli ${toolVersion}`,
    `# source: ${source}`,
    `# rosdistro: ${rosDistro}`,
    `# tags: ${tags && tags.length ? tags.join(", ") : "all"}`,
  ];
  if (packages && packages.length) lines.push(`# packages: ${packages.join(", ")}`);
  if (repository && repository.length) lines.push(`# repository: ${repository.join(", ")}`);
  if (autoware != null) {
    lines.push(
      `# autoware: ${autoware} (informational only — not a ref selector; the registry tracks one ref per repository)`,
    );
  }
  if (generatedAt != null) lines.push(`# generated_at: ${generatedAt}`);
  if (selection && selection.length) {
    lines.push("# selected packages by repository:");
    for (const [key, names] of selection) lines.push(`#   ${key}: ${names.join(", ")}`);
  }
  lines.push("# Generated file — re-run 'aw-index-cli compose …' to update; do not edit by hand.");
  return lines;
}

/**
 * Render the full `.repos` document (header comments + YAML body).
 * Mirrors `compose.render_repos`: `header.join("\n") + "\n\n" + body` (a blank
 * line separates the `#` header from the body), where the body renders
 * `{repositories: entries}` as block-style YAML whose content matches
 * `yaml.safe_dump(..., sort_keys=False, default_flow_style=False)` (see
 * `dumpBody`/`yamlScalar`).
 */
export function renderRepos(
  distribution,
  { tags = null, packages = null, repository = null, headerLines } = {},
) {
  const repositories = selectRepositories(distribution, { tags, packages, repository });
  const entries = toReposEntries(repositories);
  return headerLines.join("\n") + "\n\n" + dumpBody(entries);
}

/** Provenance string mirroring `registry.describe_source` for the network default. */
export function defaultSource(repo = DEFAULT_REPO, ref = DEFAULT_REF) {
  return `${repo}@${ref}`;
}

/**
 * One-call convenience assembling header + body exactly like the CLI's
 * `compose` handler (`cli._cmd_compose`). `packages`/`tags`/`repository` are the
 * selection filters AND are recorded in the header, matching
 * `aw-index-cli compose …`. Omit `generatedAt` for deterministic output
 * (equivalent to `--no-timestamp`).
 */
export function composeReposFile(
  distribution,
  {
    rosDistro,
    source,
    toolVersion = VERSION,
    tags = null,
    packages = null,
    repository = null,
    autoware = null,
    generatedAt = null,
  } = {},
) {
  const src = source != null ? source : defaultSource();
  const selection = selectRepositories(distribution, { tags, packages, repository }).map(
    ([key, , names]) => [key, names],
  );
  const headerLines = provenanceHeader({
    toolVersion,
    rosDistro,
    source: src,
    tags,
    packages,
    repository,
    autoware,
    generatedAt,
    selection,
  });
  return renderRepos(distribution, { tags, packages, repository, headerLines });
}

/**
 * The exact `aw-index-cli compose …` command that reproduces a selection of
 * individually-picked packages. `--packages` is space-separated (`nargs="*"`).
 */
export function composeCommand({ rosDistro, packages = [] } = {}) {
  const names = [...new Set(packages)].sort(cmp);
  const pkgArg = names.length ? ` --packages ${names.join(" ")}` : "";
  return `aw-index-cli compose --rosdistro ${rosDistro}${pkgArg} --stdout | vcs import src`;
}

// ---------------------------------------------------------------------------
// YAML emission — block-style YAML whose parsed *content* matches PyYAML 6.x
// `safe_dump(default_flow_style=False)`. The contract is content, not bytes
// (see `tests/test_conformance.py`), so the one thing that must hold is that
// every scalar round-trips to the same string: `yamlScalar` quotes a value
// exactly when a plain rendering would reload as a non-string (e.g. `1.20`,
// `on`). Formatting PyYAML does differently — escaping non-ASCII, folding long
// lines, the explicit `? key` form for long keys — is left to differ freely,
// since it reparses to the same content.
// ---------------------------------------------------------------------------

// The YAML implicit type resolvers (PyYAML SafeLoader patterns, ported verbatim
// with the re.X whitespace stripped). A plain scalar matching any of these would
// reload as a non-string, so it must be single-quoted to stay a string.
const RESOLVERS = [
  /^(?:yes|Yes|YES|no|No|NO|true|True|TRUE|false|False|FALSE|on|On|ON|off|Off|OFF)$/,
  /^(?:[-+]?0b[0-1_]+|[-+]?0[0-7_]+|[-+]?(?:0|[1-9][0-9_]*)|[-+]?0x[0-9a-fA-F_]+|[-+]?[1-9][0-9_]*(?::[0-5]?[0-9])+)$/,
  /^(?:[-+]?(?:[0-9][0-9_]*)\.[0-9_]*(?:[eE][-+][0-9]+)?|\.[0-9][0-9_]*(?:[eE][-+][0-9]+)?|[-+]?[0-9][0-9_]*(?::[0-5]?[0-9])+\.[0-9_]*|[-+]?\.(?:inf|Inf|INF)|\.(?:nan|NaN|NAN))$/,
  /^(?:~|null|Null|NULL|)$/,
  /^(?:[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]|[0-9][0-9][0-9][0-9]-[0-9][0-9]?-[0-9][0-9]?(?:[Tt]|[ \t]+)[0-9][0-9]?:[0-9][0-9]:[0-9][0-9](?:\.[0-9]*)?(?:[ \t]*(?:Z|[-+][0-9][0-9]?(?::[0-9][0-9])?))?)$/,
  /^(?:<<)$/,
  /^(?:=)$/,
];

// Characters that can never begin a plain scalar (PyYAML `analyze_scalar`).
const LEADING_INDICATORS = new Set([
  " ",
  "#",
  ",",
  "[",
  "]",
  "{",
  "}",
  "&",
  "*",
  "!",
  "|",
  ">",
  "@",
  "`",
  '"',
  "'",
  "%",
]);

// True when `s` cannot be rendered as a *plain* YAML scalar and so must be
// single-quoted to round-trip as a string. `s` is assumed non-empty.
function needsPlainQuoting(s) {
  const first = s[0];
  if (LEADING_INDICATORS.has(first)) return true;
  // '-', '?', ':' are indicators only when followed by a space or at end.
  if ((first === "-" || first === "?" || first === ":") && (s.length === 1 || s[1] === " ")) {
    return true;
  }
  // A ':' that ends the scalar (or precedes a space) is a block indicator.
  if (s[s.length - 1] === ":" || s.includes(": ")) return true;
  if (s.startsWith("---") || s.startsWith("...")) return true;
  if (s[s.length - 1] === " ") return true;
  if (s.includes(" #")) return true;
  return false;
}

/**
 * Render one scalar for block-style YAML: plain when that round-trips as the
 * same string, otherwise single-quoted (with `'` doubled). This is the one
 * subtle piece — e.g. a numeric-looking tag `1.20`, or a branch `on`/`no`, must
 * be quoted so it reloads as a string rather than a float/bool. Exported for
 * unit testing.
 */
export function yamlScalar(value) {
  const s = String(value);
  if (s === "") return "''";
  if (RESOLVERS.some((re) => re.test(s)) || needsPlainQuoting(s)) {
    return `'${s.replace(/'/g, "''")}'`;
  }
  return s;
}

// Emit one `    <name>: <scalar>` value line, always on a single line. PyYAML
// would fold a long spaced value across lines; this port does not, and both
// reparse to the same string, so content is unaffected.
function dumpField(name, value) {
  return `    ${name}: ${yamlScalar(value)}\n`;
}

// Render `{repositories: entries}` for our fixed shape: a `repositories:`
// mapping keyed by repo key, each value a 3-field mapping. Keys are always
// written inline (`key:`); PyYAML switches to the explicit `? key` form for
// empty or >=123-char keys, but that reparses to the same key.
function dumpBody(entries) {
  if (entries.size === 0) return "repositories: {}\n";
  let out = "repositories:\n";
  for (const [key, entry] of entries) {
    out += `  ${yamlScalar(key)}:\n`;
    out += dumpField("type", entry.type);
    out += dumpField("url", entry.url);
    out += dumpField("version", entry.version);
  }
  return out;
}
